### PR TITLE
Handle operators when rewriting expression trees

### PR DESCRIPTION
**Pull request by 333fred**
Currently, we're not handling binary operators in the MethodToClass rewriter, which causes user-defined operators to be missed when we lift things to closures. Fixes https://github.com/dotnet/roslyn/issues/56376. Recommend reviewing each commit separately: commit 1 nullable-enables MethodToClassRewriter, and commit 2 actually fixes the bug.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs [hunk 1]
```diff
@@ -6002,5 +6002,241 @@ .maxstack  3
...
+        static void Main()
+        {
+            GenericMethod<string>();
+        }
+
+        private static void GenericMethod<T>() where T : class
+        {
+            Func<Expression<Func<T, Wrapper<T>, bool>>> func = () => (x, y) => x == y;
+
+            func();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:6032]: >func(); [](http://example.com/codeflow?start=12&length=7)

I think we should evaluate the expression tree and observe that it calls the operator. This is applicable to other tests.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs [hunk 2]
```diff
@@ -6002,5 +6002,241 @@ .maxstack  3
...
+  IL_002f:  ldc.i4.1
+  IL_0030:  newarr     ""System.Linq.Expressions.ParameterExpression""
+  IL_0035:  dup
+  IL_0036:  ldc.i4.0
+  IL_0037:  ldloc.0
+  IL_0038:  stelem.ref
+  IL_0039:  call       ""System.Linq.Expressions.Expression<System.Func<BadImageFormatExceptionRepro.Program.Wrapper<T>, BadImageFormatExceptionRepro.Program.Wrapper<T>>> System.Linq.Expressions.Expression.Lambda<System.Func<BadImageFormatExceptionRepro.Program.Wrapper<T>, BadImageFormatExceptionRepro.Program.Wrapper<T>>>(System.Linq.Expressions.Expression, params System.Linq.Expressions.ParameterExpression[])""
+  IL_003e:  ret
+}");
+        }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:6144]: >} [](http://example.com/codeflow?start=8&length=1)

Please ensure we have similate test coverage for VB

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:6144]: >} [](http://example.com/codeflow?start=8&length=1)

It feels like we should also test user defined conversions and user defined short-circuiting operators (represented by BoundUserDefinedConditionalLogicalOperator nodes after initial binding)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:6144]: >} [](http://example.com/codeflow?start=8&length=1)

The PR description indicates that there was a problem specifically around lifting things to closure, but it appears that none of the added tests cause lifting to occur. I guess that is fine if the fix can be observed without lifting, but it would be good to have at least one test per distinct scenario using it.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:6144]: It was really about the lowering to a separate class, not the closure fields themselves. I don't think there's anything to test with the actual closure hoist itself.

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [925109942](https://github.com/dotnet/roslyn/pull/56598#issuecomment-925109942) [](http://example.com/codeflow?ancestors=925109942,925109942)

File: src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs [hunk 3]
```diff
@@ -518,16 +542,17 @@ public override BoundNode VisitFunctionPointerLoad(BoundFunctionPointerLoad node
...
         public override BoundNode VisitLoweredConditionalAccess(BoundLoweredConditionalAccess node)
         {
-            BoundExpression receiver = (BoundExpression)this.Visit(node.Receiver);
-            BoundExpression whenNotNull = (BoundExpression)this.Visit(node.WhenNotNull);
-            BoundExpression whenNullOpt = (BoundExpression)this.Visit(node.WhenNullOpt);
+            var receiver = (BoundExpression)this.Visit(node.Receiver);
+            var whenNotNull = (BoundExpression)this.Visit(node.WhenNotNull);
+            var whenNullOpt = (BoundExpression?)this.Visit(node.WhenNullOpt);
             TypeSymbol type = this.VisitType(node.Type);
             return node.Update(receiver, VisitMethodSymbol(node.HasValueMethodOpt), whenNotNull, whenNullOpt, node.Id, type);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:586]: >Debug.Assert(method is not null); [](http://example.com/codeflow?start=12&length=33)

Not sure what is the motivation for this change.

File: src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs [hunk 4]
```diff
@@ -632,10 +633,10 @@ public override BoundNode VisitReadOnlySpanFromArray(BoundReadOnlySpanFromArray
             return node.Update(operand, method, type);
         }
 
-        private static bool BaseReferenceInReceiverWasRewritten(BoundExpression originalReceiver, BoundExpression rewrittenReceiver)
+        private static bool BaseReferenceInReceiverWasRewritten([NotNullWhen(true)] BoundExpression? originalReceiver, [NotNullWhen(true)] BoundExpression? rewrittenReceiver)
         {
-            return originalReceiver != null && originalReceiver.Kind == BoundKind.BaseReference &&
-                   rewrittenReceiver != null && rewrittenReceiver.Kind != BoundKind.BaseReference;
+            return originalReceiver is { Kind: BoundKind.BaseReference } &&
+                   rewrittenReceiver is { Kind: BoundKind.BaseReference };
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs]: >rewrittenReceiver is { Kind: BoundKind.BaseReference }; [](http://example.com/codeflow?start=19&length=55)

Consider not bundling refactorings like this with product changes in the same PR. This Also doesn't look like a change related to enabling nullable annotations. It is a good thing we have tests that cought this bug.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 7)

**Comment by 333fred** [comment]
I'm going to rebase this PR now that Aleksey has reviewed it for ease of the next person to review it to separate out the nullable enable and the code change.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2). Don't forget to retarget the PR for 17.0
